### DISCUSSION TITLE
Remix should await thrown promises to see if it's a Response

### DISCUSSION THREAD

**Discussion by kentcdodds** [remix-run/remix#7043] [Proposals]
I have the following utility:

```ts
export async function logout(request: Request) {
	const cookieSession = await getSession(request.headers.get('cookie'))
	cookieSession.unset(sessionKey)
	throw redirect('/', {
		headers: { 'set-cookie': await commitSession(cookieSession) },
	})
}
```

A natural way to use it is like so:

```ts
if (inABadSpot) {
	throw logout(request)
}
```

However, if you do that, you get something similar to this:

```
GET /settings/profile/change-email 500 - - 23.605 ms

Promise {
  NodeResponse [Response] {
    size: 0,
    [Symbol(Body internals)]: {
      body: null,
      type: null,
      size: 0,
      boundary: null,
      disturbed: false,
      error: null
    },
    [Symbol(Response internals)]: {
      url: undefined,
      status: 302,
      statusText: '',
      headers: [Object],
      counter: 0,
      highWaterMark: undefined
    }
  }
}

Unable to get error message for error Promise {
  NodeResponse [Response] {
    size: 0,
    [Symbol(Body internals)]: {
      body: null,
      type: null,
      size: 0,
      boundary: null,
      disturbed: false,
      error: null
    },
    [Symbol(Response internals)]: {
      url: undefined,
      status: 302,
      statusText: '',
      headers: [Object],
      counter: 0,
      highWaterMark: undefined
    }
  }
}

GET /settings/profile/change-email 500 - - 23.004 ms

GET /settings/profile 200 - - 25.033 ms

Error: Unexpected Server Error
    at handleDataRequestRR (/Users/kentcdodds/code/epicweb-dev/web-auth/node_modules/@remix-run/server-runtime/dist/server.js:11:3525)
    at requestHandler (/Users/kentcdodds/code/epicweb-dev/web-auth/node_modules/@remix-run/server-runtime/dist/server.js:11:1320)
    at /Users/kentcdodds/code/epicweb-dev/web-auth/node_modules/@remix-run/express/dist/server.js:11:420

GET /settings/profile/change-email?_data=routes/settings+/profile.change-email 500 - - 12.864 ms
```

To fix it you have to do this:

```ts
if (inABadSpot) {
	throw await logout(request)
}
```

This is unfortunate because TypeScript can't help you with this problem since it's a `throw` and all bets are off with that.

I think it'd be great if Remix identified a thrown promise and waited for it to resolve to see whether it's a response and act accordingly.

**Comment by AlemTuzlak** [comment]
This would be awesome. This scenarios has bitten us in the behind a lot of times because we would throw a redirect like this and not setting the `await` and our site would crash. We added a jsdoc to remind you to set the await keyword but thats not really a great way to handle it

**Comment by ryanflorence** [comment]
This is kind of interesting. You're proposing to be able to throw a `Response | Promise<Response>` from a loader/action instead of just `Response` for Remix's current handling of thrown responses.

Can anybody think of some downsides of this?

**Reply by kentcdodds** [reply]
That's what I'm proposing 👍

I can't think of a reason someone would throw any other kind of promise from a loader/action. We have deferred which takes a real promise, but the promise is an object property on the response which is returned and not thrown.

So no, I don't see a downside to this.
